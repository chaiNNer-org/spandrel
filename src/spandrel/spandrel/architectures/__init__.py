"""
The package containing the implementations of all supported architectures. Not necessary for most user code.
"""
