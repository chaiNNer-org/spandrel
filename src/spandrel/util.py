"""
A module containing commonly-used functionality to implement architectures.
"""

from __future__ import annotations

import functools
import inspect
import math
from typing import Any, Literal, Mapping, Protocol, TypeVar


class KeyCondition:
    """
    A condition that checks if a state dict has the given keys.
    """

    def __init__(
        self, kind: Literal["all", "any"], keys: tuple[str | KeyCondition, ...]
    ):
        self._keys = keys
        self._kind = kind

    @staticmethod
    def has_all(*keys: str | KeyCondition) -> KeyCondition:
        return KeyCondition("all", keys)

    @staticmethod
    def has_any(*keys: str | KeyCondition) -> KeyCondition:
        return KeyCondition("any", keys)

    def __call__(self, state_dict: Mapping[str, object]) -> bool:
        def _detect(key: str | KeyCondition) -> bool:
            if isinstance(key, KeyCondition):
                return key(state_dict)
            return key in state_dict

        if self._kind == "all":
            return all(_detect(key) for key in self._keys)
        else:
            return any(_detect(key) for key in self._keys)


def get_first_seq_index(state_dict: Mapping[str, object], key_pattern: str) -> int:
    """
    Returns the maximum index `i` such that `key_pattern.format(str(i))` is in `state`.

    If no such key is in state, then `-1` is returned.

    Example:
        get_first_seq_index(state, "body.{}.weight") -> -1
        get_first_seq_index(state, "body.{}.weight") -> 3
    """
    for i in range(100):
        if key_pattern.format(str(i)) in state_dict:
            return i
    return -1


def get_seq_len(state_dict: Mapping[str, object], seq_key: str) -> int:
    """
    Returns the length of a sequence in the state dict.

    The length is detected by finding the maximum index `i` such that
    `{seq_key}.{i}.{suffix}` is in `state` for some suffix.

    Example:
        get_seq_len(state, "body") -> 5
    """
    prefix = seq_key + "."

    keys: set[int] = set()
    for k in state_dict.keys():
        if k.startswith(prefix):
            index = k[len(prefix) :].split(".", maxsplit=1)[0]
            keys.add(int(index))

    if len(keys) == 0:
        return 0
    return max(keys) + 1


def get_scale_and_output_channels(x: int, input_channels: int) -> tuple[int, int]:
    """
    Returns a scale and number of output channels such that `scale**2 * out_nc = x`.

    This is commonly used for pixelshuffel layers.
    """
    # Unfortunately, we do not have enough information to determine both the scale and
    # number output channels correctly *in general*. However, we can make some
    # assumptions to make it good enough.
    #
    # What we know:
    # - x = scale * scale * output_channels
    # - output_channels is likely equal to input_channels
    # - output_channels and input_channels is likely 1, 3, or 4
    # - scale is likely 1, 2, 4, or 8

    def is_square(n: int) -> bool:
        return math.sqrt(n) == int(math.sqrt(n))

    # just try out a few candidates and see which ones fulfill the requirements
    candidates = [input_channels, 3, 4, 1]
    for c in candidates:
        if x % c == 0 and is_square(x // c):
            return int(math.sqrt(x // c)), c

    raise AssertionError(
        f"Expected output channels to be either 1, 3, or 4."
        f" Could not find a pair (scale, out_nc) such that `scale**2 * out_nc = {x}`"
    )


def store_hyperparameters(*, extra_parameters: Mapping[str, object] = {}):
    """
    Stores the hyperparameters of a class in a `hyperparameters` attribute.
    """

    def get_arg_defaults(spec: inspect.FullArgSpec) -> dict[str, Any]:
        defaults = {}
        if spec.kwonlydefaults is not None:
            defaults = spec.kwonlydefaults

        if spec.defaults is not None:
            defaults = {
                **defaults,
                **dict(zip(spec.args[-len(spec.defaults) :], spec.defaults)),
            }

        return defaults

    class WithHyperparameters(Protocol):
        hyperparameters: dict[str, Any]

    C = TypeVar("C", bound=WithHyperparameters)

    def inner(cls: type[C]) -> type[C]:
        old_init = cls.__init__

        spec = inspect.getfullargspec(old_init)
        defaults = get_arg_defaults(spec)

        if spec.varargs is not None:
            raise UserWarning(
                "Class has *args, which is not allowed in combination with @store_hyperparameters"
            )
        if spec.varkw is not None:
            raise UserWarning(
                "Class has **kwargs, which is not allowed in combination with @store_hyperparameters"
            )

        @functools.wraps(old_init)
        def new_init(self: C, **kwargs):
            # remove extra parameters from kwargs
            for k, v in extra_parameters.items():
                if k in kwargs:
                    if kwargs[k] != v:
                        raise ValueError(
                            f"Expected hyperparameter {k} to be {v}, but got {kwargs[k]}"
                        )
                    del kwargs[k]

            self.hyperparameters = {**extra_parameters, **defaults, **kwargs}
            old_init(self, **kwargs)

        cls.__init__ = new_init
        return cls

    return inner


__all__ = [
    "KeyCondition",
    "get_first_seq_index",
    "get_seq_len",
    "get_scale_and_output_channels",
    "store_hyperparameters",
]
