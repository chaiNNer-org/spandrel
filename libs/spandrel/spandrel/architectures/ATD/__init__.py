import math

from typing_extensions import override

from spandrel.util import KeyCondition, get_pixelshuffle_params, get_seq_len

from ...__helpers.model_descriptor import (
    Architecture,
    ImageModelDescriptor,
    SizeRequirements,
    StateDict,
)
from .arch.atd_arch import ATD


class ATDArch(Architecture[ATD]):
    def __init__(
        self,
    ) -> None:
        super().__init__(
            id="ATD",
            name="ATD",
            detect=KeyCondition.has_all(
                "relative_position_index_SA",
                "conv_first.weight",
                "conv_first.bias",
                "layers.0.residual_group.td",
                "layers.0.residual_group.layers.0.sigma",
                "layers.0.residual_group.layers.0.norm1.weight",
                "layers.0.residual_group.layers.0.norm1.bias",
                "layers.0.residual_group.layers.0.norm2.weight",
                "layers.0.residual_group.layers.0.norm2.bias",
                "layers.0.residual_group.layers.0.norm3.weight",
                "layers.0.residual_group.layers.0.norm3.bias",
                "layers.0.residual_group.layers.0.wqkv.weight",
                "layers.0.residual_group.layers.0.attn_win.relative_position_bias_table",
                "layers.0.residual_group.layers.0.attn_win.proj.weight",
                "layers.0.residual_group.layers.0.attn_win.proj.bias",
                "layers.0.residual_group.layers.0.attn_atd.scale",
                "layers.0.residual_group.layers.0.attn_atd.wq.weight",
                "layers.0.residual_group.layers.0.attn_atd.wk.weight",
                "layers.0.residual_group.layers.0.attn_atd.wv.weight",
                "layers.0.residual_group.layers.0.attn_aca.logit_scale",
                "layers.0.residual_group.layers.0.attn_aca.proj.weight",
                "layers.0.residual_group.layers.0.convffn.fc1.weight",
                "layers.0.residual_group.layers.0.convffn.fc1.bias",
                "layers.0.residual_group.layers.0.convffn.dwconv.depthwise_conv.0.weight",
                "layers.0.residual_group.layers.0.convffn.dwconv.depthwise_conv.0.bias",
                "layers.0.residual_group.layers.0.convffn.fc2.weight",
                "layers.0.residual_group.layers.0.convffn.fc2.bias",
                "norm.weight",
                "norm.bias",
            ),
        )

    @override
    def load(self, state_dict: StateDict) -> ImageModelDescriptor[ATD]:
        img_size = 64  # cannot be deduced from state dict
        patch_size = 1  # cannot be deduced from state dict
        in_chans = 3
        embed_dim = 90
        depths = (6, 6, 6, 6)
        num_heads = (6, 6, 6, 6)
        window_size = 8
        category_size = 256  # cannot be deduced from state dict
        num_tokens = 64
        reducted_dim = 4
        convffn_kernel_size = 5
        mlp_ratio = 2.0
        qkv_bias = True
        ape = False
        patch_norm = True
        upscale = 1
        img_range = 1.0  # cannot be deduced from state dict
        upsampler = ""
        resi_connection = "1conv"
        norm = True

        in_chans = state_dict["conv_first.weight"].shape[1]
        embed_dim = state_dict["conv_first.weight"].shape[0]

        window_size = math.isqrt(state_dict["relative_position_index_SA"].shape[0])

        num_layers = get_seq_len(state_dict, "layers")
        depths = [6] * num_layers
        num_heads = [6] * num_layers
        for i in range(num_layers):
            depths[i] = get_seq_len(state_dict, f"layers.{i}.residual_group.layers")
            num_heads[i] = state_dict[
                f"layers.{i}.residual_group.layers.0.attn_win.relative_position_bias_table"
            ].shape[1]

        num_tokens = state_dict[
            "layers.0.residual_group.layers.0.attn_atd.scale"
        ].shape[0]
        reducted_dim = state_dict[
            "layers.0.residual_group.layers.0.attn_atd.wq.weight"
        ].shape[0]
        convffn_kernel_size = state_dict[
            "layers.0.residual_group.layers.0.convffn.dwconv.depthwise_conv.0.weight"
        ].shape[2]
        mlp_ratio = (
            state_dict["layers.0.residual_group.layers.0.convffn.fc1.weight"].shape[0]
            / embed_dim
        )
        qkv_bias = "layers.0.residual_group.layers.0.wqkv.bias" in state_dict
        ape = "absolute_pos_embed" in state_dict
        patch_norm = "patch_embed.norm.weight" in state_dict

        resi_connection = "1conv" if "layers.0.conv.weight" in state_dict else "3conv"

        if "conv_up1.weight" in state_dict:
            upsampler = "nearest+conv"
            upscale = 4
        elif "conv_before_upsample.0.weight" in state_dict:
            upsampler = "pixelshuffle"
            upscale, _ = get_pixelshuffle_params(state_dict, "upsample")
        elif "conv_last.weight" in state_dict:
            upsampler = ""
            upscale = 1
        else:
            upsampler = "pixelshuffledirect"
            upscale = math.isqrt(state_dict["upsample.0.weight"].shape[0] // in_chans)

        norm = "no_norm" not in state_dict

        is_light = upsampler == "pixelshuffledirect" and embed_dim == 48
        # use a heuristic for category_size
        category_size = 128 if is_light else 256

        tags = [f"{embed_dim}dim", f"{window_size}w", f"{category_size}cat"]
        if is_light:
            tags.insert(0, "light")

        model = ATD(
            img_size=img_size,
            patch_size=patch_size,
            in_chans=in_chans,
            embed_dim=embed_dim,
            depths=depths,
            num_heads=num_heads,
            window_size=window_size,
            category_size=category_size,
            num_tokens=num_tokens,
            reducted_dim=reducted_dim,
            convffn_kernel_size=convffn_kernel_size,
            mlp_ratio=mlp_ratio,
            qkv_bias=qkv_bias,
            ape=ape,
            patch_norm=patch_norm,
            upscale=upscale,
            img_range=img_range,
            upsampler=upsampler,
            resi_connection=resi_connection,
            norm=norm,
        )

        return ImageModelDescriptor(
            model,
            state_dict,
            architecture=self,
            purpose="Restoration" if upscale == 1 else "SR",
            tags=tags,
            supports_half=False,  # TODO: verify
            supports_bfloat16=True,
            scale=upscale,
            input_channels=in_chans,
            output_channels=in_chans,
            size_requirements=SizeRequirements(minimum=8),
        )
