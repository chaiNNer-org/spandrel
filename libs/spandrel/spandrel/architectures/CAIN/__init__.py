from __future__ import annotations

import math
from typing import Union

from typing_extensions import override

from spandrel.util import KeyCondition

from ...__helpers.model_descriptor import Architecture, ImageModelDescriptor, StateDict
from .arch.cain import CAIN
from .arch.cain_encdec import CAIN_EncDec
from .arch.cain_noca import CAIN_NoCA

_CainModel = Union[CAIN, CAIN_NoCA, CAIN_EncDec]


class CAINArch(Architecture[_CainModel]):
    def __init__(self) -> None:
        super().__init__(
            id="CAIN",
            detect=KeyCondition.has_any(
                # CAIN
                KeyCondition.has_all(
                    "encoder.interpolate.headConv.weight",
                    "encoder.interpolate.headConv.bias",
                    "encoder.interpolate.body.0.body.0.body.0.conv.weight",
                    "encoder.interpolate.body.0.body.0.body.2.conv.weight",
                    "encoder.interpolate.body.0.body.0.body.3.conv_du.0.weight",
                    "encoder.interpolate.body.0.body.1.body.0.conv.weight",
                    "encoder.interpolate.body.0.body.1.body.3.conv_du.2.weight",
                    "encoder.interpolate.body.0.body.3.body.2.conv.weight",
                    "encoder.interpolate.body.3.body.8.body.0.conv.weight",
                    "encoder.interpolate.body.4.body.5.body.3.conv_du.2.weight",
                    "encoder.interpolate.tailConv.weight",
                    "encoder.interpolate.tailConv.bias",
                ),
                # CAIN_NoCA
                KeyCondition.has_all(
                    "encoder.interpolate.headConv.weight",
                    "encoder.interpolate.headConv.bias",
                    "encoder.interpolate.body.0.body.0.body.0.conv.weight",
                    "encoder.interpolate.body.0.body.0.body.2.conv.weight",
                    "encoder.interpolate.body.0.body.1.body.0.conv.weight",
                    "encoder.interpolate.body.0.body.3.body.2.conv.weight",
                    "encoder.interpolate.body.3.body.8.body.0.conv.weight",
                    "encoder.interpolate.tailConv.weight",
                    "encoder.interpolate.tailConv.bias",
                ),
                # CAIN_EncDec
                KeyCondition.has_all(
                    "encoder.body.0.conv.weight",
                    "encoder.body.6.conv.weight",
                    "encoder.interpolate.headConv.weight",
                    "encoder.interpolate.headConv.bias",
                    "encoder.interpolate.body.0.body.0.body.0.conv.weight",
                    "encoder.interpolate.body.0.body.0.body.2.conv.weight",
                    "encoder.interpolate.body.0.body.0.body.3.conv_du.0.weight",
                    "encoder.interpolate.body.0.body.1.body.0.conv.weight",
                    "encoder.interpolate.body.0.body.1.body.3.conv_du.2.weight",
                    "encoder.interpolate.body.0.body.3.body.2.conv.weight",
                    "encoder.interpolate.body.3.body.8.body.0.conv.weight",
                    "encoder.interpolate.body.4.body.5.body.3.conv_du.2.weight",
                    "encoder.interpolate.tailConv.weight",
                    "encoder.interpolate.tailConv.bias",
                    "decoder.body.1.body.0.conv.weight",
                    "decoder.body.1.body.2.conv.bias",
                    "decoder.body.3.body.0.conv.weight",
                    "decoder.body.3.body.2.conv.bias",
                    "decoder.body.5.body.0.conv.weight",
                    "decoder.body.5.body.2.conv.bias",
                ),
            ),
        )

    @override
    def load(self, state_dict: StateDict) -> ImageModelDescriptor[_CainModel]:
        tags: list[str] = []

        if "encoder.body.0.conv.weight" in state_dict:
            start_filters = 32
            up_mode = "shuffle"

            start_filters = state_dict["decoder.body.5.body.0.conv.weight"].shape[0]

            if "decoder.body.0.upconv.0.conv.weight" in state_dict:
                up_mode = "shuffle"
            elif "decoder.body.0.upconv.weight" in state_dict:
                up_mode = "transpose"
            else:
                up_mode = "direct"

            model = CAIN_EncDec(
                start_filters=start_filters,
                up_mode=up_mode,
            )
            tags.extend(("EncDec", f"{start_filters}sf"))
        else:
            depth = 3

            # detect
            # n_feats = 3 * (4**depth)
            n_feats = state_dict["encoder.interpolate.headConv.weight"].shape[1]
            depth = math.isqrt(math.isqrt(n_feats // 3))

            if (
                "encoder.interpolate.body.0.body.0.body.3.conv_du.0.weight"
                in state_dict
            ):
                model = CAIN(
                    depth=depth,
                )
            else:
                model = CAIN_NoCA(
                    depth=depth,
                )
                tags.append("NoCA")

            tags.append(f"{depth}depth")

        return ImageModelDescriptor(
            model,
            state_dict,
            architecture=self,
            purpose="Restoration",
            tags=tags,
            supports_half=True,
            supports_bfloat16=True,
            scale=1,
            input_channels=3,
            output_channels=3,
        )
