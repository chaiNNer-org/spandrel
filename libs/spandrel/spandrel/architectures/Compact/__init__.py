from typing_extensions import override

from spandrel.util import KeyCondition, get_scale_and_output_channels, get_seq_len

from ...__helpers.model_descriptor import Architecture, ImageModelDescriptor, StateDict
from .arch.SRVGG import SRVGGNetCompact


class CompactArch(Architecture[SRVGGNetCompact]):
    def __init__(
        self,
    ) -> None:
        super().__init__(
            id="Compact",
            name="RealESRGAN Compact",
            detect=KeyCondition.has_all(
                "body.0.weight",
                "body.1.weight",
            ),
        )

    @override
    def load(self, state_dict: StateDict) -> ImageModelDescriptor[SRVGGNetCompact]:
        state = state_dict

        highest_num = get_seq_len(state, "body") - 1

        in_nc = state["body.0.weight"].shape[1]
        num_feat = state["body.0.weight"].shape[0]
        num_conv = (highest_num - 2) // 2

        pixelshuffle_shape = state[f"body.{highest_num}.bias"].shape[0]
        scale, out_nc = get_scale_and_output_channels(pixelshuffle_shape, in_nc)

        model = SRVGGNetCompact(
            num_in_ch=in_nc,
            num_out_ch=out_nc,
            num_feat=num_feat,
            num_conv=num_conv,
            upscale=scale,
        )

        return ImageModelDescriptor(
            model,
            state,
            architecture=self,
            purpose="Restoration" if scale == 1 else "SR",
            tags=[f"{num_feat}nf", f"{num_conv}nc"],
            supports_half=True,
            supports_bfloat16=True,
            scale=scale,
            input_channels=in_nc,
            output_channels=out_nc,
        )
