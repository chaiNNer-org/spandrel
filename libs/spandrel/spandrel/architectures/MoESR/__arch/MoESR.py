import math
from typing import Literal

import torch
from torch import nn
from torch.nn import functional as F  # noqa: N812
from torch.nn.init import trunc_normal_

from spandrel.architectures.__arch_helpers.dysample import DySample
from spandrel.util import store_hyperparameters

SampleMods = Literal[
    "conv", "pixelshuffledirect", "pixelshuffle", "nearest+conv", "dysample"
]


class UniUpsample(nn.Sequential):
    def __init__(
        self,
        upsample: SampleMods,
        scale: int = 2,
        in_dim: int = 64,
        out_dim: int = 3,
        mid_dim: int = 64,  # Only pixelshuffle
        group: int = 4,  # Only DySample
    ) -> None:
        m = []

        if scale == 1 or upsample == "conv":
            m.append(nn.Conv2d(in_dim, out_dim, 3, 1, 1))
        elif upsample == "pixelshuffledirect":
            m.extend(
                [nn.Conv2d(in_dim, out_dim * scale**2, 3, 1, 1), nn.PixelShuffle(scale)]
            )
        elif upsample == "pixelshuffle":
            m.extend([nn.Conv2d(in_dim, mid_dim, 3, 1, 1), nn.LeakyReLU(inplace=True)])
            if (scale & (scale - 1)) == 0:  # scale = 2^n
                for _ in range(int(math.log2(scale))):
                    m.extend(
                        [nn.Conv2d(mid_dim, 4 * mid_dim, 3, 1, 1), nn.PixelShuffle(2)]
                    )
            elif scale == 3:
                m.extend([nn.Conv2d(mid_dim, 9 * mid_dim, 3, 1, 1), nn.PixelShuffle(3)])
            else:
                raise ValueError(
                    f"scale {scale} is not supported. Supported scales: 2^n and 3."
                )
            m.append(nn.Conv2d(mid_dim, out_dim, 3, 1, 1))
        elif upsample == "nearest+conv":
            if (scale & (scale - 1)) == 0:
                for _ in range(int(math.log2(scale))):
                    m.extend(
                        (
                            nn.Conv2d(in_dim, in_dim, 3, 1, 1),
                            nn.Upsample(scale_factor=2),
                            nn.LeakyReLU(negative_slope=0.2, inplace=True),
                        )
                    )
                m.extend(
                    (
                        nn.Conv2d(in_dim, in_dim, 3, 1, 1),
                        nn.LeakyReLU(negative_slope=0.2, inplace=True),
                    )
                )
            elif scale == 3:
                m.extend(
                    (
                        nn.Conv2d(in_dim, in_dim, 3, 1, 1),
                        nn.Upsample(scale_factor=scale),
                        nn.LeakyReLU(negative_slope=0.2, inplace=True),
                        nn.Conv2d(in_dim, in_dim, 3, 1, 1),
                        nn.LeakyReLU(negative_slope=0.2, inplace=True),
                    )
                )
            else:
                raise ValueError(
                    f"scale {scale} is not supported. Supported scales: 2^n and 3."
                )
            m.append(nn.Conv2d(in_dim, out_dim, 3, 1, 1))
        elif upsample == "dysample":
            m.append(DySample(in_dim, out_dim, scale, group))
        else:
            raise ValueError(
                f"An invalid Upsample was selected. Please choose one of {SampleMods}"
            )
        super().__init__(*m)

        self.register_buffer(
            "MetaUpsample",
            torch.tensor(
                [
                    1,  # Block version, if you change something, please number from the end so that you can distinguish between authorized changes and third parties
                    list(SampleMods.__args__).index(upsample),  # pyright: ignore  # UpSample method index
                    scale,
                    in_dim,
                    out_dim,
                    mid_dim,
                    group,
                ],
                dtype=torch.uint8,
            ),
        )


class LayerNorm(nn.Module):
    def __init__(self, dim: int = 64, eps: float = 1e-6) -> None:
        super().__init__()
        self.weight = nn.Parameter(torch.ones(dim))
        self.bias = nn.Parameter(torch.zeros(dim))
        self.eps = eps

    def forward(self, x):
        u = x.mean(1, keepdim=True)
        s = (x - u).pow(2).mean(1, keepdim=True)
        x = (x - u) / torch.sqrt(s + self.eps)
        return self.weight[:, None, None] * x + self.bias[:, None, None]


class InceptionDWConv2d(nn.Module):
    """Inception depthweise convolution"""

    def __init__(
        self, in_channels, square_kernel_size=3, band_kernel_size=11, branch_ratio=0.125
    ) -> None:
        super().__init__()

        gc = int(in_channels * branch_ratio)  # channel numbers of a convolution branch
        self.dwconv_hw = nn.Conv2d(
            gc, gc, square_kernel_size, padding=square_kernel_size // 2, groups=gc
        )
        self.dwconv_w = nn.Conv2d(
            gc,
            gc,
            kernel_size=(1, band_kernel_size),
            padding=(0, band_kernel_size // 2),
            groups=gc,
        )
        self.dwconv_h = nn.Conv2d(
            gc,
            gc,
            kernel_size=(band_kernel_size, 1),
            padding=(band_kernel_size // 2, 0),
            groups=gc,
        )
        self.split_indexes = [in_channels - 3 * gc, gc, gc, gc]

    def forward(self, x):
        x_id, x_hw, x_w, x_h = torch.split(x, self.split_indexes, dim=1)
        return torch.cat(
            (x_id, self.dwconv_hw(x_hw), self.dwconv_w(x_w), self.dwconv_h(x_h)),
            dim=1,
        )


class GatedCNNBlock(nn.Module):
    r"""
    modernized mambaout main unit
    https://github.com/yuweihao/MambaOut/blob/main/models/mambaout.py#L119
    """

    def __init__(
        self,
        dim: int = 64,
        expansion_ratio: float = 8 / 3,
        conv_ratio: float = 1.0,
    ) -> None:
        super().__init__()
        self.norm = LayerNorm(dim)
        hidden = int(expansion_ratio * dim)
        self.fc1 = nn.Conv2d(dim, hidden * 2, 3, 1, 1)

        self.act = nn.Mish()
        conv_channels = int(conv_ratio * dim)
        self.split_indices = [hidden, hidden - conv_channels, conv_channels]

        self.conv = InceptionDWConv2d(conv_channels)
        self.fc2 = nn.Conv2d(hidden, dim, 3, 1, 1)
        self.gamma = nn.Parameter(torch.ones([1, dim, 1, 1]), requires_grad=True)
        self.apply(self._init_weights)

    @staticmethod
    def _init_weights(m) -> None:
        if isinstance(m, (nn.Conv2d, nn.Linear)):
            trunc_normal_(m.weight, std=0.02)
            if m.bias is not None:
                nn.init.constant_(m.bias, 0)

    def forward(self, x):
        shortcut = x
        x = self.norm(x)
        g, i, c = torch.split(self.fc1(x), self.split_indices, dim=1)
        c = self.conv(c)
        x = self.act(self.fc2(self.act(g) * torch.cat((i, c), dim=1)))
        return (x * self.gamma) + shortcut


class MSG(nn.Module):
    def __init__(self, dim, expansion_msg=1.5) -> None:
        super().__init__()
        self.down = nn.Sequential(
            nn.Conv2d(dim, dim // 4, 3, 1, 1),
            nn.PixelUnshuffle(2),
            nn.LeakyReLU(0.1, True),
        )
        self.gated = nn.Sequential(
            *[GatedCNNBlock(dim, expansion_ratio=expansion_msg) for _ in range(3)]
        )
        self.up = nn.Sequential(
            nn.Conv2d(dim, dim * 4, 3, 1, 1),
            nn.PixelShuffle(2),
            nn.LeakyReLU(0.1, True),
        )

    def forward(self, x):
        out = self.down(x)
        out = self.gated(out)
        return self.up(out) + x


class Blocks(nn.Module):
    def __init__(
        self,
        dim: int = 64,
        blocks: int = 4,
        expansion_factor: float = 1.5,
        expansion_msg: float = 1.5,
    ) -> None:
        super().__init__()
        self.blocks = nn.Sequential(
            *[GatedCNNBlock(dim, expansion_factor) for _ in range(blocks)]
        )
        self.msg = MSG(dim, expansion_msg)

    def forward(self, x):
        return self.msg(self.blocks(x))


@store_hyperparameters()
class MoESR(nn.Module):
    """Mamba out Excitation Super-Resolution"""

    hyperparameters = {}

    def __init__(
        self,
        *,
        in_ch: int = 3,
        out_ch: int = 3,
        scale: int = 4,
        dim: int = 64,
        n_blocks: int = 9,
        n_block: int = 4,
        expansion_factor: float = 2.65625,
        expansion_msg: float = 1.5,
        upsampler: SampleMods = "pixelshuffledirect",
        upsample_dim: int = 64,
    ) -> None:
        super().__init__()
        if upsampler == "conv":
            scale = 1
        self.scale = scale
        self.in_to_dim = nn.Conv2d(in_ch, dim, 3, 1, 1)
        self.blocks = nn.Sequential(
            *[
                Blocks(dim, n_block, expansion_factor, expansion_msg)
                for _ in range(n_blocks)
            ]
        )
        self.upscale = UniUpsample(upsampler, scale, dim, out_ch, upsample_dim)

    def check_img_size(self, x, resolution):
        h, w = resolution
        scaled_size = 2
        mod_pad_h = (scaled_size - h % scaled_size) % scaled_size
        mod_pad_w = (scaled_size - w % scaled_size) % scaled_size
        return F.pad(x, (0, mod_pad_w, 0, mod_pad_h), "reflect")

    def forward(self, x):
        _b, _c, h, w = x.shape
        x = self.check_img_size(x, (h, w))
        x = self.in_to_dim(x)
        x = self.blocks(x) + x
        return self.upscale(x)[:, :, : h * self.scale, : w * self.scale]
