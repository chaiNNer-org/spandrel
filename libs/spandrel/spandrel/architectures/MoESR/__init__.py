from spandrel.__helpers.size_req import SizeRequirements
from spandrel.util import KeyCondition, get_seq_len

from ...__helpers.model_descriptor import Architecture, ImageModelDescriptor, StateDict
from .__arch.MoESR import MoESR, SampleMods


class MoESRArch(Architecture[MoESR]):
    def __init__(self):
        super().__init__(
            id="MoESR",
            # https://github.com/rewaifu/resselt/blob/989dca4fa6d4c32ef09adc948ff26963fee807d5/resselt/archs/moesr/__init__.py#L13
            detect=KeyCondition.has_all(
                "in_to_dim.weight",
                "in_to_dim.bias",
                "blocks.0.blocks.0.gamma",
                "blocks.0.blocks.0.norm.weight",
                "blocks.0.blocks.0.norm.bias",
                "blocks.0.blocks.0.fc1.weight",
                "blocks.0.blocks.0.fc1.bias",
                "blocks.0.blocks.0.conv.dwconv_hw.weight",
                "blocks.0.blocks.0.conv.dwconv_hw.bias",
                "blocks.0.blocks.0.conv.dwconv_w.weight",
                "blocks.0.blocks.0.conv.dwconv_w.bias",
                "blocks.0.blocks.0.conv.dwconv_h.weight",
                "blocks.0.blocks.0.conv.dwconv_h.bias",
                "blocks.0.blocks.0.fc2.weight",
                "blocks.0.blocks.0.fc2.bias",
                "upscale.MetaUpsample",
            ),
        )

    def load(self, state_dict: StateDict) -> ImageModelDescriptor:
        upsample: list[SampleMods] = [
            "conv",
            "pixelshuffledirect",
            "pixelshuffle",
            "nearest+conv",
            "dysample",
        ]
        dim, in_ch = state_dict["in_to_dim.weight"].shape[:2]
        n_blocks = get_seq_len(state_dict, "blocks")
        n_block = get_seq_len(state_dict, "blocks.0.blocks")
        expansion_factor_shape = state_dict["blocks.0.blocks.0.fc1.weight"].shape
        expansion_factor = (expansion_factor_shape[0] / expansion_factor_shape[1]) / 2
        expansion_msg_shape = state_dict["blocks.0.msg.gated.0.fc1.weight"].shape
        expansion_msg = (expansion_msg_shape[0] / expansion_msg_shape[1]) / 2
        _, index, scale, _, out_ch, upsample_dim, _ = state_dict["upscale.MetaUpsample"]
        upsampler = upsample[int(index)]

        model = MoESR(
            in_ch=in_ch,
            out_ch=int(out_ch),
            scale=int(scale),
            n_blocks=n_blocks,
            n_block=n_block,
            dim=dim,
            expansion_factor=expansion_factor,
            expansion_msg=expansion_msg,
            upsampler=upsampler,
            upsample_dim=int(upsample_dim),
        )

        return ImageModelDescriptor(
            model,
            state_dict,
            architecture=self,
            purpose="Restoration" if scale == 1 else "SR",
            tags=[f"{dim}dim", f"{n_blocks}nbs", f"{n_block}nb", upsampler],
            supports_half=True,
            supports_bfloat16=True,
            scale=int(scale),
            input_channels=in_ch,
            output_channels=int(out_ch),
            size_requirements=SizeRequirements(minimum=2, multiple_of=1, square=False),
        )


__all__ = ["MoESRArch", "MoESR"]
