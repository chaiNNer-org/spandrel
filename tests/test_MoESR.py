from spandrel.architectures.MoESR import MoESR, MoESRArch

from .util import (
    ModelFile,
    TestImage,
    assert_image_inference,
    assert_loads_correctly,
    assert_size_requirements,
    disallowed_props,
)


def test_load():
    assert_loads_correctly(
        MoESRArch(),
        lambda: MoESR(),
        lambda: MoESR(in_ch=1, out_ch=1),
        lambda: MoESR(n_blocks=10),
        lambda: MoESR(n_block=5),
        lambda: MoESR(dim=32),
        lambda: MoESR(scale=2),
        lambda: MoESR(scale=1, upsampler="conv"),
        lambda: MoESR(upsampler="pixelshuffle"),
        lambda: MoESR(upsampler="dysample"),
        lambda: MoESR(expansion_factor=2.0),
        lambda: MoESR(expansion_msg=1.0),
        lambda: MoESR(upsample_dim=32),
    )


def test_size_requirements():
    file = ModelFile.from_url(
        "https://github.com/the-database/traiNNer-redux/releases/download/pretrained-models/4x_DF2K_MoESR_500k.safetensors",
        name="4x_DF2K_MoESR_500k.safetensors",
    )
    assert_size_requirements(file.load_model())


def test_4x_df2k_moesr(snapshot):
    file = ModelFile.from_url(
        "https://github.com/the-database/traiNNer-redux/releases/download/pretrained-models/4x_DF2K_MoESR_500k.safetensors",
        name="4x_DF2K_MoESR_500k.safetensors",
    )
    model = file.load_model()
    assert model == snapshot(exclude=disallowed_props)
    assert isinstance(model.model, MoESR)
    assert_image_inference(
        file,
        model,
        [TestImage.SR_16, TestImage.SR_32, TestImage.SR_64],
    )
