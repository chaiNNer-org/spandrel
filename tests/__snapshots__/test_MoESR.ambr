# serializer version: 1
# name: test_4x_df2k_moesr
  ImageModelDescriptor(
    architecture=MoESRArch(
      id='MoESR',
      name='MoESR',
    ),
    input_channels=3,
    output_channels=3,
    purpose='SR',
    scale=4,
    size_requirements=SizeRequirements(minimum=2, multiple_of=1, square=False),
    supports_bfloat16=True,
    supports_half=True,
    tags=list([
      '64dim',
      '9nbs',
      '4nb',
      'pixelshuffledirect',
    ]),
    tiling=<ModelTiling.SUPPORTED: 1>,
  )
# ---
